src/ports/postgres/modules/dbscan/dbscan.py_in

```diff
@@ -26,81 +26,297 @@ from utilities.utilities import add_postfix
+                rt_counts_dict[i['__leaf_id__']] = int(i['count'])
+            rt_counts_list = []
+            for i in sorted(rt_counts_dict):
+                rt_counts_list.append(rt_counts_dict[i])
+
+            leaf_id_start = pow(2,depth)-1
+
+            find_core_points_table = unique_string(desp='find_core_points_table')
+            rt_edge_table = unique_string(desp='rt_edge_table')
+            rt_core_points_table = unique_string(desp='rt_core_points_table')
+            border_core_points_table = unique_string(desp='border_core_points_table')
+            border_edge_table = unique_string(desp='border_edge_table')
+            plpy.execute("DROP TABLE IF EXISTS {0}, {1}, {2}, {3}, {4}".format(
+                find_core_points_table, rt_edge_table, rt_core_points_table,
+                border_core_points_table, border_edge_table))
+
+            sql = """
+            CREATE TABLE {find_core_points_table} AS
+            SELECT __leaf_id__,
+                   {schema_madlib}.find_core_points( {id_column},
+                                               {expr_point}::DOUBLE PRECISION[],
+                                               {eps},
+                                               {min_samples},
+                                               '{metric}',
+                                               ARRAY{rt_counts_list},
+                                               __leaf_id__
+                                               )
+            FROM {cur_source_table} GROUP BY __leaf_id__
+            {dist_leaf_sql}
+            """.format(**locals())
+            plpy.execute(sql)
+
+            sql = """
+            CREATE TABLE {rt_edge_table} AS
+            SELECT (unpacked_2d).src AS __src__, (unpacked_2d).dest AS __dest__
+            FROM (
+                SELECT {schema_madlib}.unpack_2d(find_core_points) AS unpacked_2d
+                FROM {find_core_points_table}
+                ) q1
+            WHERE (unpacked_2d).src NOT IN (SELECT {id_column} FROM {border_table1})
+            {dist_src_sql}
+            """.format(**locals())
+            plpy.execute(sql)
+
+            sql = """
+                CREATE TABLE {rt_core_points_table} AS
+                SELECT DISTINCT(__src__) AS {id_column} FROM {rt_edge_table}
+                """.format(**locals())
+            plpy.execute(sql)
+
+            # # Start border
+            sql = """
+                CREATE TABLE {border_edge_table} AS
+                SELECT __src__, __dest__ FROM (
+                    SELECT  __t1__.{id_column} AS __src__,
+                            __t2__.{id_column} AS __dest__,
+                            {schema_madlib}.{metric}(
+                                __t1__.{expr_point}, __t2__.{expr_point}) AS __dist__
+                    FROM {border_table1} AS __t1__, {border_table2} AS __t2__)q1
+                WHERE __dist__ < {eps}
+                """.format(**locals())
+            plpy.execute(sql)
```

Comment:
This query is an O(B^2) operation, where B is the number of boundary points.  Same as brute force algorithm, but with N replaced by B.  At first I thought this wouldn't be too bad, because it seems like B would be very small compared to N.  Unfortunately, this reasoning only works well for a low number of dimensions.  As the # of dimensions gets higher, the hypervolume of the boundary is not that different from the overall hypervolume.
Here's a calculation of the runtime... I think this is the bottleneck for this whole algorithm.

D = dimension of space
N = number of points
S = number of segments
eps = epsilon

Hypervolume of space containing all points = V = L^D
Hypervolume of boundaries = O(eps * L^(D-1))
Hypervolume of boundaries between boundaries = O(eps^2 * L^(D-2))
...

Avg density of points = n = N/V = N/L^D

Points in boundary regions = O(eps*n*L^(D-1)) = (eps/L)*O(N)

Current runtime:  (1/S)*(eps/L)^2*O(N)^2

The goal should be able to do this in  O(N/S*log(N)).  I think we can achieve this if we re-run the `find_core_points()` AGGREGATE on the boundary points, then on the boundaries of the boundaries, etc. until we reach 1D boundaries.